Python/pylifecycle.c

```diff
@@ -1251,17 +1251,18 @@ finalize_interp_clear(PyThreadState *tstate)
 {
     int is_main_interp = _Py_IsMainInterpreter(tstate);
 
-    /* bpo-36854: Explicitly clear the codec registry
-       and trigger a GC collection */
+    /* bpo-36854: Explicitly clear the codec registry */
     PyInterpreterState *interp = tstate->interp;
     Py_CLEAR(interp->codec_search_path);
     Py_CLEAR(interp->codec_search_cache);
     Py_CLEAR(interp->codec_error_registry);
-    _PyGC_CollectNoFail();
 
     /* Clear interpreter state and all thread states */
     PyInterpreterState_Clear(tstate->interp);
 
+    /* Trigger a GC collection */
+    _PyGC_CollectNoFail();
```

Comment:
I'm not fully comfortable with triggering a GC collection after PyInterpreterState_Clear(). If a finalizer tries to access an interpreter state or a Python thread state, it will likely get an inconsistent object or an error. I'm not sure that finalizers handle well every possible error cases. For example, what happens if a finalizer tries to import a module at this point?

May we could try to split PyInterpreterState_Clear() in two functions or more. Or we could modify _PyGC_CollectNoFail() to not call finalizers. But what's the point of a GC collection in this case?

Maybe I'm too pessimistic and it will be fine.

I suggest to not hold this PR. Subinterpreters is a work in progress anyway.

**Maybe a temporary solution is to not call _PyGC_CollectNoFail() in the main interpreter to mimick Python 3.8 (to reduce the risk of regression), but only call it in subinterpreters.**